server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeQueue.java

```diff
@@ -0,0 +1,144 @@
+        if (ThreadPoolMergeScheduler.USE_THREAD_POOL_MERGE_SCHEDULER_SETTING.get(settings)) {
+            return new ThreadPoolMergeQueue(threadPool);
+        } else {
+            return null;
+        }
+    }
+
+    private ThreadPoolMergeQueue(ThreadPool threadPool) {
+        this.executorService = threadPool.executor(ThreadPool.Names.MERGE);
+        this.maxConcurrentMerges = threadPool.info(ThreadPool.Names.MERGE).getMax();
+    }
+
+    void submitMergeTask(MergeTask mergeTask) {
+        enqueueMergeTask(mergeTask);
+        executeNextMergeTask();
+    }
+
+    void enqueueMergeTask(MergeTask mergeTask) {
+        queuedMergeTasks.add(mergeTask);
+        if (mergeTask.supportsIOThrottling()) {
+            activeIOThrottledMergeTasksCount.incrementAndGet();
+            maybeUpdateTargetMBPerSec();
+        }
+    }
+
+    private void executeNextMergeTask() {
+        executorService.execute(() -> {
+            // one such task always executes a SINGLE merge task; this is important for merge queue statistics
+            while (true) {
+                MergeTask smallestMergeTask;
+                try {
+                    // will block if there are backlogged merges until they're enqueued again
+                    smallestMergeTask = queuedMergeTasks.take();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    return;
+                }
+                if (smallestMergeTask.runNowOrBacklog()) {
+                    runMergeTask(smallestMergeTask);
+                    // one runnable one merge task
+                    return;
+                }
+            }
+        });
+    }
+
+    private void runMergeTask(MergeTask mergeTask) {
+        assert mergeTask.isRunning() == false;
+        boolean added = currentlyRunningMergeTasks.add(mergeTask);
+        assert added : "starting merge task [" + mergeTask + "] registered as already running";
+        try {
+            if (mergeTask.supportsIOThrottling()) {
+                mergeTask.setIORateLimit(targetMBPerSec.get());
+            }
+            mergeTask.run();
+        } finally {
+            boolean removed = currentlyRunningMergeTasks.remove(mergeTask);
+            assert removed : "completed merge task [" + mergeTask + "] not registered as running";
+            if (mergeTask.supportsIOThrottling()) {
+                activeIOThrottledMergeTasksCount.decrementAndGet();
+                maybeUpdateTargetMBPerSec();
+            }
+        }
+    }
+
+    private void maybeUpdateTargetMBPerSec() {
+        double currentTargetMBPerSec = targetMBPerSec.get();
+        double newTargetMBPerSec = newTargetMBPerSec(currentTargetMBPerSec, activeIOThrottledMergeTasksCount.get(), maxConcurrentMerges);
+        if (currentTargetMBPerSec != newTargetMBPerSec && targetMBPerSec.compareAndSet(currentTargetMBPerSec, newTargetMBPerSec)) {
```

Comment:
I think we should use `targetMBPerSec.updateAndGet()` to update this, handling concurrency without missing adjustment of the target value.